### ISSUE TITLE
Improve PYI051 docs to indicate that it is an opinionated stylistic rule

### ISSUE THREAD

**Issue by puc9** [astral-sh/ruff#23134]
### Summary

The issue here is that annotation `Literal['a','b'] | str` is not equivalent with `Literal['a','b']`.
The first allows an editor to suggest `'a'` and `'b'` as possible values but allows any other `str` value as well.

The second annotation is very clear that only `'a'` and `'b'` are valid values.

```python
# this function fails PYI051
def fails_PYI051(lit: Literal['a','b'] | str) -> None:
    pass

# this function passes PYI051
def passes_PYI051(lit: Literal['a','b']) -> None:
    pass

fails_PYI051('a') # this call passes mypy, as expected
fails_PYI051('c') # this call passes mypy, and it is expected to pass, according to the annotations

passes_PYI051('a') # this call passes mypy, as expected
passes_PYI051('c') # this call fails mypy, as expected
```
mypy playground: https://mypy-play.net/?mypy=latest&python=3.12&gist=6ba9044ebdb38ddcf5cc170524021161
*(I couldn't figure out how to add the rule for PYI051 to the playground)*

**Comment by tar-ang-2004** [comment]
@puc9 Heyy......I’ll take a look and try to reproduce this locally. I want to confirm whether the rule behavior is intentional or a bug in the implementation. I’ll report back with findings.

**Comment by tar-ang-2004** [comment]
**Update:** I was able to reproduce this locally. Ruff flags Literal[...] | str because each literal is treated as redundant once it’s unioned with the broader str type, so the annotation effectively collapses to str. That matches the current rule behavior. At the same time, I see the concern that the literal still communicates intent even if it’s technically redundant. 

Should PYI051 consider this more of a style preference, or is the strict behavior intentional?

**Comment by puc9** [comment]
Thanks a lot for taking a look!

### TL;DR:

IMHO the current rule behavior is not desirable and not a style preference only.

To resolve the rule violation the user would need to drop part of the information that is in the annotation. A `Literal['a', 'b'] | str` annotation is a poor man's extendable `enum` and they help tremenduosly when using unfamiliar API and it makes it easier to write the code.

I can drop the rule, which is what I do now by using `# noqa` just in the places I encounter this issue, but I'm not yet sure that it is a completely useless rule so I don't want to just simply disable it, or at least not for now.

Also, clearly not a big issue, but thougth I should report it nonetheless.

### Details:

I'm a typing enthusiast, not a specialist, so take my opinion with a grain of salt.

In my opinion the `PYI051` rule (and other similar `PYI` rules) are half-right that the annotation contains redundant information from a strict type checking perspective.

Not all information is redundant in this case, it provides an editor the ability to offer code completion with the list of currenlty known possible values, even when that list isn't fixed, so a different value can be used. Think of the many libraries that have as a main selling point the fact that they use `typing` to improve editor experience.

This applies to other `PYI` rules as well, for example `PYI041` that complains that `def fct(val: int | float):` is redundant and should be collapsed to `def fct(val: float):`. For a new Python user (that's myself) that relies heavily on the typing annotations to understand the API they try to use, both of there rules lead to a loss of information.

For example I had no idea that some numeric values are special cased in the typing spec and that I can pass an int to a float param, especially since this does not apply to all numeric values (I'm looking at you, `Decimal`, so hard to use).

For example `ruff` autofixes:

```python
def fct(val: int | float) -> int | float:  # noqa: PYI041
    return 0
```

into

```python
def fct(val: float) -> int | float:
    return 0
```

**Comment by tar-ang-2004** [comment]
Thanks for the detailed explanation.

I think that makes your concern much clearer. I see what you mean about literals carrying intent even when they’re technically redundant from a strict type perspective. In cases like `Literal[...] | str`, the annotation does more than just describe the final type; it also helps readers and editor tooling understand expected values.

I think the interesting question is how Ruff wants to balance strict normalization vs preserving that semantic information. I’m curious what the maintainers think about that tradeoff.

**Comment by ntBre** [comment]
Thank you for the report and @tar-ang-2004 for looking into it as well!

I'd be curious to get @AlexWaygood's thoughts here. I think the rule is operating as intended, but I can see the point about editor integration. Based on the `Redundant code` category from the [flake8-pyi docs](https://github.com/PyCQA/flake8-pyi/blob/main/ERRORCODES.md#Y051), I don't think you'd be missing out on any correctness issues by disabling the rule.

**Comment by AlexWaygood** [comment]
This rule is indeed operating "as designed". It's an opinionated rule that might not suit everybody; it's disabled by default, and I think it should remain so. The premise of the rule is that we found a lot of typeshed contributors were adding redundant unions to their annotations entirely by accident; we wanted to add a lint rule to the flake8-pyi plugin to help flag that.

However, the rule dates from a time when we only really considered type checkers (not language servers/IDEs) to be the primary users of typeshed (and stubs in general). Over the last few years, we've changed our stance on that at typeshed; we now consider language servers/IDEs to be just as important a consumer of typeshed's stubs as command-line type checkers. Given that change in stance -- which I think reflects a change in community opinion more broadly across the ecosystem -- this rule is probably not one that I'd be excited to add to flake8-pyi or Ruff in 2025. We could possibly consider deprecating it from both flake8-pyi and Ruff.

**Comment by tar-ang-2004** [comment]
Thanks for the context. That makes sense. 

If this ends up moving toward deprecation, I’d be happy to help with `docs` or a `deprecation note` so users can understand the reasoning and migration path.

**Comment by tjkuson** [comment]
I agree the rule is opinionated, but I like it as-is and would be sad to see it deprecated. It can catch discrepancies between what is written and the developer's intent and (in my opinion) makes type annotations easier to read.

I get that the redundant literals can aid some developer's IDEs. However, I think that it's equally resolved by `typing.overload` which seems more appropriate for describing multiple valid argument types. `typing.overload` is also more powerful than `typing.Literal` approach in that it can describe relationships with other types in the signature. I've used this pattern before to nudge users in the right direction without adding noise to the types of the implementation itself.

**Comment by AlexWaygood** [comment]
Hmm, then maybe the best resolution here would just be to update the docs for this rule so that it's clearer that it's an opinionated lint that may not suit everybody's needs?

**Comment by Avasam** [comment]
This is definitely a pedantic rule, one that if you enable, you have to expect you'll have to noqa in a few places to clarify your intent that "you know the union is redundant, but it helps a human reader\* understand possible intended values"
\* or a LLM I guess...

It can also help IDEs with autocompletion, whilst still allowing arbitrary values (it's a popular trick in the TS ecosystem)

---

Once rule recategorization happens, this, and rules about string length, probably should not be enabled by default in a "type correctness / best practices" category.

Edit: Right, very long strings do hurt editor visibility
> And they're also likely to not be useful for IDE users -- you generally wouldn't want a very long string showing up in a tooltip when you hover over a function.

**Comment by AlexWaygood** [comment]
Oh, I think the rules about string length are different. Part of the reason we added those is that we found that we went "too far" in many cases when we initally added default values to tyepshed. We accidentally added default values in situations where it wasn't appropriate to do so (because, for example, the default value was dynamically computed depending on the platform, Python version or even the location of the user's Python installation). You can see https://github.com/python/typeshed/pull/9572/files for some examples of that kind of thing being fixed.

Very long strings in stubs (e.g. basedir: str = "/Users/jelle/.pyenv/versions/3.11.1/lib/python3.11" was a default value we added to typeshed at one point by mistake) are more likely to be incorrect in this way. And they're also likely to not be useful for IDE users -- you generally wouldn't want a very long string showing up in a tooltip when you hover over a function.

I'd lean towards those ones being enabled by default. But this discussion about other rules is off-topic for this issue about PYI051 :-)

**Comment by tjkuson** [comment]
> Once rule recategorization happens, this [...] probably should not be enabled by default in a "type correctness / best practices" category.

To be clear, I like the rule as-is both

1. for stylistic reasons (opinionated and possibly pedantic, I concede), and
2. to nudge users towards `typing.overload` which is a more explicit† and more powerful way to achieve the same thing, where invalid hints are more likely to be caught by the type-checker than `typing.Literal` pattern (I'd argue that this is a 'better practice').

Whether that qualifies the rule for inclusion in one category over another is probably off-topic, but I think if we're updating the documentation, a mention of (2) could be beneficial?

† a redundant `typing.Literal` union is often a mistake

**Comment by AlexWaygood** [comment]
As I said, I think we should update the docs here, but I'm not sure I'd add a recommendation for `typing.overload` to these docs... I'd generally avoid `overload` where possible, as it's pretty verbose and the rules around overloads can get pretty complex. I'm not saying it's _bad_ to prefer using `@overload`, but the motivation doesn't quite seem strong enough for me to warrant mentioning it in the docs for PYI051.

**Comment by tar-ang-2004** [comment]
I can draft a docs change that:

- explicitly marks PYI051 as an opinionated stylistic rule
- explains the original motivation behind the rule
- clarifies when disabling it may be reasonable

I’ll focus on updating the rule description without expanding its recommendations.
